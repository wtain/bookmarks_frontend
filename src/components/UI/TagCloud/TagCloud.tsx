
import { useEffect, useState } from "react";
import ITagsRepository from "../../../domain/repository/tags/ITagsRepository";
import Tag from "../TagsEditor/Tag";

interface Props {
    tagsRepository: ITagsRepository;
    onTagClick?: (tag: string) => void;
    onTagSelected?: (tag: string) => void;
    onTagDeSelected?: (tag: string) => void;
    onClearTagsClick?: () => void;
}

const TagCloud = (props: Props) => {

    let [tags, setTags] = useState(
        [] as string[]
    )

    let [, setLoading] = useState(false);
    
    const [selectedTags, setSelectedTags] = useState<string[]>([]); // new Set() ?

    const loadTags = async(success: (tags: string[]) => void, error: (e: any) => void) => {
        try {
            const tags = await props.tagsRepository.getTags();
            success(tags)
        } catch (e) {
            error(e)
        }
    }

    useEffect(() => {
        setLoading(true);
        loadTags((tags: string[]) => {
            setTags(tags)
            setLoading(false);
        }, 
            (e) => {
                setLoading(false);
            })
    }, [])

    return (
        <>
            <Tag tag={{ name: "❌", id: "" }} onClick={() => {
                if (props.onClearTagsClick) {
                    props.onClearTagsClick!();
                }
                setSelectedTags([]);
            }} />
            {tags.map(tag => <Tag key={tag} tag={{ name: tag, id: "" }}
                isSelected={selectedTags.includes(tag)}
                onClick={() => {
                    if (props.onTagClick) {
                        props.onTagClick!(tag);      
                    }
                    if (selectedTags.includes(tag)) {
                        //selectedTags.delete(tag);
                        setSelectedTags(selectedTags.filter(v => v !== tag));
                        if (props.onTagDeSelected) {
                            props.onTagDeSelected!(tag);
                        }
                    }
                    else {
                        //selectedTags.add(tag);
                        setSelectedTags([...selectedTags, tag]);
                        if (props.onTagSelected) {
                            props.onTagSelected!(tag);
                        }
                    }
                }} />)}
        </>
    )
}

export default TagCloud;