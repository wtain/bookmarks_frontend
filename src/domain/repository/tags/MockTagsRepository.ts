import ReactHelpers from '../../../utils/ReactHelpers';
import ITagsRepository from './ITagsRepository';


class MockTagsRepository implements ITagsRepository {
  async getTags(): Promise<string[]> {
    await ReactHelpers.delay(300);
    return ["tag1", "tag2", "tag3"];
  }
  
}

export default MockTagsRepository;