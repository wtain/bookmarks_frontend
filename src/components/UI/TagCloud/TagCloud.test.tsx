import renderer from 'react-test-renderer';
import MockTagsRepository from '../../../domain/repository/tags/MockTagsRepository';
import { BrowserRouter } from 'react-router-dom';
import TagCloud from './TagCloud';

it('TagCloud renders correctly', () => {
  const tagsRepository = new MockTagsRepository();

  const tree = renderer
    .create(
      <BrowserRouter>
        <TagCloud tagsRepository={tagsRepository} />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});