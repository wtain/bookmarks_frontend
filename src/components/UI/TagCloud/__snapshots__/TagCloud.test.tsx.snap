// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagCloud renders correctly 1`] = `
<span
  className="tag"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  (Clear)
</span>
`;
